### PR TITLE
add testing blog post

### PR DESCRIPTION
**Pull request by JanPietrzyk**

### CODE REVIEW

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+When we set of to write the B2B-Suite we as a team set a few technical goals we wanted to achieve. One of them was to embrace automatic testing at its fullest and harness the potential implied by it:
+
+* Allow refactoring even a day before release
+* Help instill trust into the stability of the application for our customers
+* Create a good application architecture that feels about right for the complexity introduced
+
+## The Application
+
+First we take a look at a fairly standard application design. We need this to have a shared base for the comparision of the different testing strategies. This model omits the framework and data structures which are for our not essential for our evaluation.
```
Comment (teiling88) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: which are for *us* not essential for our evaluation

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+We have a controller, that uses a domain service and two auxiliary services to handle requests. The services themselves have dependencies on other services or storage implementations.
+
+<img src="/blog/img/advantages-of-integration-testing/standard-application.svg" />
+
+So how should we test this? Lt's discuss some strategies:
```
Comment (teiling88) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: L**e**t's

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 1 above)
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: ... wich are not essential for our evaluation

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 2 above)
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: Let's

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+
+## Testing Strategy: Functional testing only
+
+Many projects out there use acceptance tests as their only way of automated testing. So why is it a bad idea? One word: Performance! Functional tests use a real browser to check the application. Even high performance applications have response times of at least 50ms, excluding the JavaScript execution time you are quickly at 500ms or more - for a **single page**. So let's math with an average of 250ms per page:
```
Comment (teiling88) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: ggf. is **this** a

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 2 above)
Comment (Christian-Rades) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: Missing 'e'

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+* Gives you 99% certainty that the application works correctly.
+* Adds minimal code to maintain.
+
+**Bad:**
+
+* Takes a long time even for small applications
+* Does not help developers
+* Does nothing to your application architecture, but add a few css classes
+* Usually very fragile, small changes tend to have an unreasonably big impact (you change a buttons position the login breaks)
```
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: .. position **and** the login breaks

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+**Verdict:** Run!
+
+## Testing Strategy: The Pyramid
+
+There is a testing pyramid favored by people like Robert C. Martin, it looks like this:
+
+<img src="/blog/img/advantages-of-integration-testing/unit-test-pyramid.svg" />
+
+The Testing Pyramid puts the focus on **unit tests** proposes to write a few integration tests and if really necessary very few functional test. So for our example application this looks like this:
```
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: Our example application looks like:

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+
+<img src="/blog/img/advantages-of-integration-testing/standard-application-with-pyramid.svg" />
+
+So what do we have here? We have a Unit Test for each class of the application. Each dependency is mocked so we can test the code of a single service in absolute isolation. This is the core of your test suite: The unit tests! Now we need to integrate in a few places, I personally would propose that `MyService` and `ContextProvider` should be integrated with their subsidiary services, so we add this:
+
+<img src="/blog/img/advantages-of-integration-testing/standard-application-with-pyramid-integration.svg" />
+
+These two integration tests that integrate the central services with there storage. Last but not least we have a small functional test for our controller.
```
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: ... **their** storage

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+
+## Testing Strategy: The Diamond
+
+There is a third strategy, that puts all the effort into Integration tests. This looks something like that:
+
+<img src="/blog/img/advantages-of-integration-testing/test-diamond.svg" />
+
+I have color coded the application diagram to show you how we interpreted the diamond:
+
+* **Green:** No dependencies, a test becomes outomatically a **unit test**
```
Comment (teiling88) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: **a**utomatically

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 8]
```diff
@@ -0,0 +1,282 @@
...
+
+<img src="/blog/img/advantages-of-integration-testing/standard-application-with-pyramid-integration.svg" />
+
+These two integration tests that integrate the central services with there storage. Last but not least we have a small functional test for our controller.
+
+<img src="/blog/img/advantages-of-integration-testing/standard-application-with-pyramid-functional.svg" />
+
+Pyramid successfully implemented!
+
+But this did not work for us. Since we use minimal Integration tests and mock all output of the dependencies we quickly found ourselves in a place where the test suite did not produce certain edge cases corectly. Mocks are simply a lie in your system. Let's take a closer look at `MyServiceTest`:
```
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: .. all **outputs**

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 5 above)
Comment (Christian-Rades) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: Maybe better : "A Testing Pyramid _that_ puts the focus on unit tests . . .  few functional tests"

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+
+* **Green:** No dependencies, a test becomes outomatically a **unit test**
+* **Yellow:** Has dependencies (either on other classes, or on infrastructure) must be an **integration test**
+* **Blue:** Is front facing so it must be a **functional test**
+
+<img src="/blog/img/advantages-of-integration-testing/application-diamond-testing.svg" />
+
+By applying this simple ruleset we still have a one to one parity of test classes to production classes. This means that for each bug or feature there is an easy entry point in our test suite. And the test suite is therefore useful for developers and gets actively supported. If bugs are introduced during development there is usually a test that shows it immediately. If a contract between objects is changed this change is immediately available in the suite and in many cases even leads to a failing test. (depending on the test quality: see [Mutation Testing by my colleague Thomas Eiling](https://developers.shopware.com/blog/2017/08/24/mutation-testing/)).
+
+So now lets take a look at our `MyServiceTest` the way it looks like in integration testing:
```
Comment (teiling88) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: let**'**s

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 10]
```diff
@@ -0,0 +1,282 @@
...
+And then gets changed to this:
+
+```php
+public function isValidResponseCode(int $code): bool
+{
+    return in_array($code, [200, 300, 202, 303]);
+}
+```
+
+Boom, a mock will never automatically produce this result, but for all means it will keep the test suite green. There are certain other problems with mocks, so... Time for a excurse:
```
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: **an** excurse

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+
+* Almost doubles the amount of code in your project
+* Fixture data will never show 100% production state
+* Will also not give you a 100% certainty that everything works
+
+**Verdict:** We us it!
+
+## A word on TDD
+
+I personally am a huge fan of TDD and practice it almost always when developing or exploring software. This practice is not lost here. Just because you have to implement stuff that you are using does not reduce the amount of freedom you have in the 30 second cycle of test driven development. On the contrary - at least for me - it tends to create more realistic assumptions on the scope of classes. By doing almost only integration testing all my tests look and feel the same. The usual friction that occurs when combining multiple unit tested classes is by far reduced because the integration assumptions all - by definition -  hold up against the real implementations.
```
Comment (teiling88) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: ggf. I am personally

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 12]
```diff
@@ -0,0 +1,282 @@
...
+}
+```
+
+Boom, a mock will never automatically produce this result, but for all means it will keep the test suite green. There are certain other problems with mocks, so... Time for a excurse:
+
+----
+
+## Excurse: Mocking
+
+All I really can say about mocking is this: it is a pain in the ass! There are currently a few contenders for mocking in PHPUnit tests and they all work internally the same way and are kind of awful. So lets say we have a class called `Something` that has a method called `getResponse` and we want it to return `'foo'` for our test, how, would we accomplish this through mocking?
```
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: how without commata

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 11 above)
Comment (teiling88) [reply source/blog/_posts/2017-11-17-on-testing-strategies.md]: double white space between - and hold up against the real im

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 13]
```diff
@@ -0,0 +1,282 @@
...
+
+## Conclusion
+
+The Diamond seems to be the best solution to the testing strategy problem. In fact we have been and still are developing the B2B-Suite with that strategy in mind and have seen quite good results with it. Currently our complete suite takes around 3 minutes to execute which is still passable for a quick overview. None of our tests depend on it's predecessor and can all be executed in isolation (which makes Mutation testing possible in the first place), so during development we usually single out a few tests to run repeatedly.
```
Comment (Christian-Rades) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: > None of our tests depend on it's predecessor 
-> None of our tests depend on _its_ predecessor

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 14]
```diff
@@ -0,0 +1,282 @@
...
+
+The issue of software architecture comes a little short in this blog post, let me tell you why: It is a problem where tests only are a part of the solution! As you could read in a [previous post I published](https://developers.shopware.com/blog/2016/12/05/large-scale-plugin-architecture/) where I used the actor model to create a plugin architecture there are numerous ways that need to be taken into account when creating a sustainable system architecture. Testing will help you a lot with *friction* and *cohesion* (as introduced by Robert C. Martin) but not help you derive a greater truth from this. This is a place where [DDD](https://en.wikipedia.org/wiki/Domain-driven_design) seems to provide the better tools.
+
+So all in all, at least to us this is the best overall approach to developing stable software, with the least tradeoffs and the most benefits. Thanks for reading and as always feel free to get in touch and discuss with me!
```
Comment (emmer91) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: to **develop** stable software

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 11 above)
Comment (JanPietrzyk) [reply source/blog/_posts/2017-11-17-on-testing-strategies.md]: google sagt geht auch so rum

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 3 above)
Comment (JanPietrzyk) [reply source/blog/_posts/2017-11-17-on-testing-strategies.md]: math find ich ja lustiger

File: source/blog/_posts/2017-11-17-on-testing-strategies.md (hunk 5 above)
Comment (JanPietrzyk) [reply source/blog/_posts/2017-11-17-on-testing-strategies.md]: das ist n anderes Diagramm, deshalb sieht es für die Applikation so aus:

File: source/blog/_posts/2017-11-17-on-testing-strategies.md [hunk 15]
```diff
@@ -0,0 +1,282 @@
...
+
+* **Green:** No dependencies, a test becomes automatically a **unit test**
+* **Yellow:** Has dependencies (either on other classes, or on infrastructure) must be an **integration test**
+* **Blue:** Is front facing so it must be a **functional test**
+
+<img src="/blog/img/advantages-of-integration-testing/application-diamond-testing.svg" />
+
+By applying this simple ruleset we still have a one to one parity of test classes to production classes. This means that for each bug or feature there is an easy entry point in our test suite. And the test suite is therefore useful for developers and gets actively supported. If bugs are introduced during development there is usually a test that shows it immediately. If a contract between objects is changed this change is immediately available in the suite and in many cases even leads to a failing test. (depending on the test quality: see [Mutation Testing by my colleague Thomas Eiling](https://developers.shopware.com/blog/2017/08/24/mutation-testing/)).
+
+So now lets take a look at our `MyServiceTest` the way it looks like in integration testing:
```
Comment (mitelg) [inline source/blog/_posts/2017-11-17-on-testing-strategies.md]: let's

**Comment by mitelg** [comment]
good read :+1: 
one little thing @teiling88 already mentioned :wink:
